### PR TITLE
#11 - courses in semester

### LINKED ISSUE #11: Courses in semester management
This will be more complex than typical CRUD.

We need:
* [ ] a connection between semesters and courses, eg. course "Projektowanie i programowanie obiektowe I - L" is conducted in "semestr zimowy 2023/24"
* [ ] every course in semester should have multiple groups, eg. "s2INF1(1)"
* [ ] every group should have multiple students assigned

Current solution is a little different, because there's different view for all groups, but I find it little unhandy:

obraz

obraz

### PR DESCRIPTION
**Pull request by kamilpiech97**
This should close #11.

Added the ability to create semester courses and groups with students for them. Added tests.

Index view of courses in semester:
image

Create view:
image

Edit view:
image

Managing of the groups:

[Screencast from 07-10-23 15:40:47.webm](https://github.com/blumilksoftware/keating/assets/29115024/eacdb833-baa6-4a21-8fe0-72900a56c8c9)

### CODE REVIEW

File: app/Http/Controllers/Dashboard/GroupStudentController.php [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+            "students" => $students,
+            "availableStudents" => $availableStudents,
+            "search" => $searchText,
+            "total" => $group->students->count(),
+        ]);
+    }
+
+    public function store(Request $request, CourseSemester $course, Group $group): RedirectResponse
+    {
+        $group->students()->sync($request->get("student"), false);
```
Comment (krzysztofrewak) [inline app/Http/Controllers/Dashboard/GroupStudentController.php:48]: Please add argument name here for this `false`.

File: database/migrations/2023_10_05_193906_create_groups_table.php [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+return new class() extends Migration {
+    public function up(): void
+    {
+        Schema::create("groups", function (Blueprint $table): void {
+            $table->ulid("id")->primary();
+            $table->string("name");
+            $table->string("form");
+            $table->string("course_semester_id")->nullable();
+            $table->foreign("course_semester_id")->references("id")->on("course_semester")->onDelete("set null");
```
Comment (krzysztofrewak) [inline database/migrations/2023_10_05_193906_create_groups_table.php:17]:
```suggestion
            $table->foreign("course_semester_id")->references("id")->on("course_semester")->onDelete("SET NULL");
```

As we discussed recently.

File: resources/js/Layouts/DashboardLayout.vue [hunk 3]
```diff
@@ -40,7 +40,7 @@ const navigation = [
     title: 'Uczelnia',
     elements: [
       { name: 'Studenci', href: '/dashboard/students', icon: UsersIcon, current: false },
-      { name: 'Kursy w semestrze', href: '#', icon: BookmarkSquareIcon, current: false },
+      { name: 'Kursy w semestrze', href: '/dashboard/course-semester', icon: BookmarkSquareIcon, current: false },
```
Comment (krzysztofrewak) [inline resources/js/Layouts/DashboardLayout.vue:43]:
```suggestion
      { name: 'Kursy w semestrze', href: '/dashboard/semester-courses', icon: BookmarkSquareIcon, current: false },
```

File: resources/js/Pages/Dashboard/CourseSemester/Index.vue [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+            <TableHeader class="w-1/6">
+              ID
+            </TableHeader>
+            <TableHeader class="w-1/6">
+              Kurs
+            </TableHeader>
+            <TableHeader class="w-1/6">
+              Semestr
+            </TableHeader>
+            <TableHeader />
```
Comment (krzysztofrewak) [inline resources/js/Pages/Dashboard/CourseSemester/Index.vue:59]: I would add here number of assigned group indicator.

File: resources/js/Pages/Dashboard/CourseSemester/Show.vue [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+import { ref } from 'vue'
+
+const props = defineProps({
+  course: Object,
+  groups: Array,
+  studyForms: Array,
+})
+const showModal = ref(false)
+const groupToDeleteId = ref(0)
+const groupToEdit = ref(0)
```
Comment (krzysztofrewak) [inline resources/js/Pages/Dashboard/CourseSemester/Show.vue:25]: Why `0` and not `null`?

**Review by krzysztofrewak — CHANGES_REQUESTED** [review]

File: resources/js/Pages/Dashboard/CourseSemester/Show.vue (hunk 5 above)
Comment (kamilpiech97) [reply resources/js/Pages/Dashboard/CourseSemester/Show.vue]: fixed

File: resources/js/Pages/Dashboard/CourseSemester/Index.vue (hunk 4 above)
Comment (kamilpiech97) [reply resources/js/Pages/Dashboard/CourseSemester/Index.vue]: added

File: app/Http/Controllers/Dashboard/GroupStudentController.php (hunk 1 above)
Comment (kamilpiech97) [reply app/Http/Controllers/Dashboard/GroupStudentController.php]: added

**Review by krzysztofrewak — APPROVED** [review]

**Pull request merged**